src/layers/content_cache.rs

```diff
@@ -0,0 +1,286 @@
+///
+/// # Notes
+///
+/// This layer only maintains its own states. Users should care about the cache
+/// consistency by themselves. For example, in the following situations, users
+/// could get out-dated metadata cache:
+///
+/// - Users have operations on underlying operator directly.
+/// - Other nodes have operations on underlying storage directly.
+/// - Concurrent read/write/delete on the same path.
+///
+/// To make sure content cache consistent across the cluster, please make sure
+/// all nodes in the cluster use the same cache services like redis or tikv.
+///
+/// # Examples
+///
+/// ```
+/// use anyhow::Result;
+/// use opendal::layers::ContentCacheLayer;
+/// use opendal::services::memory;
+/// use opendal::Operator;
+/// use opendal::Scheme;
+///
+/// let _ = Operator::from_env(Scheme::Fs)
+///     .expect("must init")
+///     .layer(ContentCacheLayer::new(
+///         memory::Builder::default().build().expect("must init"),
+///     ));
+/// ```
+#[derive(Debug, Clone)]
+pub struct ContentCacheLayer {
+    cache: Arc<dyn Accessor>,
+}
+
+impl ContentCacheLayer {
+    /// Create a new metadata cache layer.
+    pub fn new(acc: impl Accessor + 'static) -> Self {
+        Self {
+            cache: Arc::new(acc),
+        }
+    }
+}
+
+impl Layer for ContentCacheLayer {
+    fn layer(&self, inner: Arc<dyn Accessor>) -> Arc<dyn Accessor> {
+        Arc::new(ContentCacheAccessor {
+            cache: self.cache.clone(),
+            inner,
+        })
+    }
+}
+
+#[derive(Debug, Clone)]
+struct ContentCacheAccessor {
+    cache: Arc<dyn Accessor>,
+    inner: Arc<dyn Accessor>,
+}
+
+#[async_trait]
+impl Accessor for ContentCacheAccessor {
+    fn metadata(&self) -> AccessorMetadata {
+        self.inner.metadata()
+    }
+
+    async fn create(&self, path: &str, args: OpCreate) -> Result<()> {
+        self.cache.delete(path, OpDelete::new()).await?;
+        self.inner.create(path, args).await
+    }
+
+    async fn read(&self, path: &str, args: OpRead) -> Result<BytesReader> {
+        match self.cache.read(path, args.clone()).await {
+            Ok(r) => Ok(r),
+            Err(err) if err.kind() == ErrorKind::NotFound => {
+                let r = self.inner.read(path, args).await?;
+                let meta = self.inner.stat(path, OpStat::new()).await?;
+                if meta.mode().is_file() {
+                    let size = meta.content_length();
+                    let reader = self.inner.read(path, OpRead::new(..)).await?;
+                    self.cache.write(path, OpWrite::new(size), reader).await?;
+                }
+                Ok(r)
+            }
+            Err(err) => Err(err),
+        }
+    }
+
+    async fn write(&self, path: &str, args: OpWrite, r: BytesReader) -> Result<u64> {
+        self.cache.delete(path, OpDelete::new()).await?;
+        self.inner.write(path, args, r).await
+    }
+
+    async fn stat(&self, path: &str, args: OpStat) -> Result<ObjectMetadata> {
```

Comment:
Only the objectmeta is needed here, but the cache write is still done on failure, this is in view of the fact that we may use this data next.